### PR TITLE
:bug: Location-agnostic sign release

### LINKED ISSUE #240: Signing should respect the ARTIFACTS_TO_PUBLISH variable
The code signing should respect the `ARTIFACTS_TO_PUBLISH` variable. If the variable contains the `checksums.txt` file, that file should be updated.

Current code expects the artifacts to live in project root.

https://github.com/knative/hack/blob/cf1a127cfdcd88b1d74fec195089a2090178a596/release.sh#L316-L359

/kind bug

### PR DESCRIPTION
**Pull request by cardil**
# Changes

- :bug: Location-agnostic sign release

/kind bug

Fixes #240
Related to https://github.com/knative-sandbox/homebrew-kn-plugins/issues/134

### CODE REVIEW

File: release.sh
Comment (kvmware) [inline release.sh:324]: Shouldn't block the merge, but I prefer and find it easier to read the `for x in blah` loop over the `while do done < input` syntax.

**Comment by kvmware** [comment]
Awesome work on adding the tests :+1: 
Got my LGTM

File: release.sh
Comment (cardil) [reply release.sh:324]: https://www.shellcheck.net/wiki/SC2013

Shell is hard.

Comment (kvmware) [reply release.sh:324]: Yes it is :(

Shellcheck folks are more bash nerds than me for sure so I will mark as resolved

Comment (cardil) [reply release.sh:324]: BTW: We should have shellcheck linter running before merge on this repo.

Comment (kvmware) [reply release.sh:324]: 100% https://github.com/knative/hack/issues/269

**Comment by cardil** [comment]
/hold

I'd say to wait for @upodroid as this is mostly his code.

**Review by upodroid — APPROVED** [review]
LGTM

We can test this by bumping the deps and rerunning the nightly. The final artifacts will be in the nightly GCS bucket

File: release.sh
Comment (cardil) [inline release.sh:354]:
```suggestion
      pushd "$(dirname "$file")" >/dev/null
      sha256sum "$(basename "$file")" >> "${checksums}"
      popd >/dev/null
```

File: release.sh
Comment (cardil) [inline release.sh:385]:
```suggestion
        pushd "$(dirname "$file")" >/dev/null
        sha256sum "$(basename "$file")" >> "${checksums}"
        popd >/dev/null
```

**Review by cardil — COMMENTED** [review]
This needs to be adjusted

File: release.sh
Comment (cardil) [reply release.sh]: 0f24a18 fixes this

Comment (evankanderson) [reply release.sh]: Does this run into an issue if we have two files with the same name in subdirectories?  I think I'd rather `cd`/`pwd` into the root release directory and then use relative paths from there.

Also wondering how you feel about the `(cd ...; ...)` construct vs `pwd ...; ...; popd`

Comment (cardil) [reply release.sh]: The checksum should not have paths because we host it on GH release page, and it should be valid in that regard. It doesn't make sense to preserve full paths, as that makes validation very problematic for end users.

You can see such validation here: https://github.com/knative-sandbox/homebrew-kn-plugins/issues/134#issuecomment-1419171040

The `(cd ...; ...)` work similar to `pushd; ...; ...; popd`, although opening a subshell is tricky as you could lose changes done withing subshell (people don't expect sub-shells)

**Comment by cardil** [comment]
Infra failure

/test integration-tests

**Comment by cardil** [comment]
/test integration-tests

Infra failure: GKE cluster failed to boot

File: release.sh
Comment (cardil) [reply release.sh:372]: I don't know to what you are referring exactly. Is it https://www.shellcheck.net/wiki/SC2013 ?

It shows couple valid options on how to iterate lines from a file/pipe/stdin. All of them are equally valid.

The problem is people keep using the invalid iteration patterns. Habits from other programming languages often makes writing shell code harder, as bash is unique in many ways.

File: release.sh
Comment (cardil) [reply release.sh:366]: I don't see any glob-equal anywhere. Could you explain it exactly?

File: release.sh
Comment (cardil) [reply release.sh:382]: I don't follow, again. There's no comparison here.

This line clears the checksums file content. The `true` built-in has no stdout, so redirecting it to the file, clears its contents.

Comment (cardil) [reply release.sh:382]: What could have been done better here, is to not clear the file entirely, but replace just the signed files.

We could make such amend in a follow up, IMHO.

File: release.sh
Comment (evankanderson) [reply release.sh:372]: I meant that this format:

```
grep -v '^ *#' < file | while IFS= read -r line
do
  echo "Line: $line"
done
```

Seems clearer to me, because the set of things you're looping over is at the start.

The linked docs suggest that the form using process substitution (`<(...)`) "avoids a subshell", but I'm not sure I understand why that would be true.  In any case, it feels like having the set of things to loop over at the beginning of the loop is a bit more natural.

File: release.sh
Comment (evankanderson) [reply release.sh:366]: On line 333, you wrote:

```
    if [[ "${file}" == *"checksums.txt" ]]; then
```

I wondered if that form would be better than `echo | grep`

Comment (evankanderson) [reply release.sh:366]: Oops, this was supposed to be on line 368:

```
      if echo "$file" | grep -q "darwin"; then
```

File: release.sh
Comment (evankanderson) [reply release.sh:382]: It looks like GitHub made a mess of the line numbers on my comments, let me find the right line.

File: release.sh
Comment (cardil) [reply release.sh:372]: Both forms are equally valid. Matter of preference, I guess.

Giving things to iterate over before the loop, can be very cloudy, especially if the command is long.

Passing it to `done` is clear to me, and avoids long lines. Feeding data to commands with `<` is quite common in shell. No point, IMHO, of making bash to look like other programming languages, it isn't similar.

Your proposal has only one minor downside. Because it uses pipe, it may fail silently without stopping the script (if `pipefail` option isn't set ofc).

File: release.sh
Comment (cardil) [reply release.sh:366]: Yes, it may be written as

```bash
if [[ "$file" == *"darwin"* ]]; then
```

IDK if it's worth changing though. This form is way less known, and it's very error prone.

For example, someone could rewrite it to:

```bash
if [[ "$file" == "*darwin*" ]]; then
```

Which means completely different thing, and will not work.

But, going back to your question: I'm not seeing much benefit in a rewrite, but we definitely could do it if you feel like it.

**Pull request merged**
